rfd/0002-custom-terraform-registry.md

```diff
@@ -0,0 +1,185 @@
+
+The existing Houston distribution tool at [get.gravitational.com](https://get.gravitational.com/)
+is not suitable for distributing the files indexed by the registry. Houston has
+_very_ particular requirements about the naming of the files it serves (e.g.
+all files are expected to be `*.tar.gz` compressed tarballs - or at least
+_named_ as such) and, as it stands, will not serve the files we need it to.
+
+We _could_ modify Houston to serve the required files, but as the service is
+already slated for retirement, that does not seem to be a good use of time
+& money.
+
+### Building provider packages
+
+As far as the registry protocol is concerned, the entire package for a given
+provider consists of:
+
+* The actual binary distribution zipfile
+* A checksum file containing the SHA256 of the zipfile (formatted as per
+ `sha256sum` tool)
+* A signature file containing a binary, detached GPG signature for
+  checksum file.
+
+The Teleport provider release process currently produces a compressed tarball
+by default, and this should still be considered the primary artefact produced
+by the build, at least as long as Houston is expected to serve downloads of
+the Terraform provider.
+
+By analogy with the main Teleport release system, the registry-compatible
+provider package will be created during release promotion, with a build task
+that will:
+
+1. For each release tarball (one each for multiple release platforms):
+    1. repack the tarball into an appropriately formatted & named zipfile
+    2. generate the checksum file from the zipfile
+    3. generate the signature file by signing the checksum file
+2. Upload all generated files to the AWS bucket that backs the registry
+3. Update the registry index to include the newly released providers (see
+   _Populating the registry_, below)
+
+I expect this to be implemented as a small Go program, rather than a shell
+script. Using small Go programs for build tasks has shown itself to be a
+useful, flexible and above all _legible_ tool for scripting build tasks in
+the mainline Teleport repository, and we should continue that pattern in
+`teleport-plugins`.
+
+#### Build Security issues
+
+This process will require that the `teleport-plugins` promotion process will
+need new secrets:
+
+1. Read access to a signing key, to sign the package (see note below)
+2. Write access to the AWS bucket storing the signed packages for distribution
+
+_**Note:**_ We currently expect to use the RPM signing key as the Terraform
+signing key as it already exists and is already used for a similar purpose.
+That said, there is no technical reason why it _must_ be the RPM signing
+key. As long as the private key used to sign the package agrees with the a
+public key exposed via the registry, the signature should be deemed valid
+by Terraform.
+
+### Populating the registry
+
+Populating the registry is a two-part problem:
+
+First, we must create a new, back-filled registry containing all of the
```

Comment:
Do we need all previously published providers? Maybe publish latest v8, latest v9 and then keep publishing new versions as we're releasing them?